src/main.cairo

```diff
@@ -0,0 +1,158 @@
+        address: ContractAddress,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct DomainResolvingUpdate {
+        #[key]
+        domain: Span<felt252>,
+        field: felt252,
+        target_addr: ContractAddress,
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        admin_address: ContractAddress,
+        starknetid_address: ContractAddress,
+        naming_address: ContractAddress,
+        pub_key: felt252
+    ) {
+        self.identity_contract.write(starknetid_address);
+        self.naming_contract.write(naming_address);
+        self.admin.write(admin_address);
+        self.server_pub_key.write(pub_key);
+    }
+
+    #[external(v0)]
+    impl SolSubdomainImpl of ISolSubdomain<ContractState> {
+        fn claim(
+            ref self: ContractState, name: felt252, sig: (felt252, felt252), max_validity: u64,
+        ) {
+            assert(get_block_timestamp() < max_validity.try_into().unwrap(), 'Signature expired');
+
+            let caller = get_caller_address();
+            let message_hash: felt252 = hash::LegacyHash::hash(
+                hash::LegacyHash::hash(hash::LegacyHash::hash('sol subdomain', max_validity), name),
+                get_caller_address()
+            );
+
+            let public_key = self.server_pub_key.read();
+            let (sig0, sig1) = sig;
+            let is_valid = check_ecdsa_signature(message_hash, public_key, sig0, sig1);
+            assert(is_valid, 'Invalid signature');
+
+            self.name_owners.write(name, caller);
+
+            self
+                .emit(
+                    Event::DomainClaimed(
+                        DomainClaimed { domain: array![name, 16434].span(), address: caller, }
+                    )
+                )
+        }
+
+        fn was_claimed(self: @ContractState, name: felt252) -> ContractAddress {
+            self.name_owners.read(name)
+        }
+
+        fn were_claimed(self: @ContractState, names: Span<felt252>) -> Array<ContractAddress> {
+            let mut result: Array<ContractAddress> = array![];
+            let mut names = names;
+            loop {
+                if names.len() == 0 {
+                    break;
+                }
+                let name = names.pop_front().unwrap();
+                result.append(self.name_owners.read(*name));
+            };
+            result
+        }
+
+        fn set_resolving(
+            ref self: ContractState, domain: felt252, field: felt252, new_target: ContractAddress
+        ) {
+            let caller = get_caller_address();
+            assert(self.name_owners.read(domain) == caller, 'Caller not owner of domain');
+            self.resolving_mapping.write(domain, new_target);
+
+            self
+                .emit(
+                    Event::DomainResolvingUpdate(
+                        DomainResolvingUpdate {
+                            domain: array![domain, 16434].span(), field, target_addr: new_target,
```

Comment:
Same suggestion here, but this one is more important because this is actually the one I will index for resolving